### PR TITLE
Adds support for Salus OTA updates

### PR DESCRIPTION
**Pull request by PedroLamas**
This will add support for Salus OTA updates (only enabled for the SP600 smart plug)

I've added a dependency to [tar-stream](https://www.npmjs.com/package/tar-stream) as the update files are tar archives.

**Note 1:** while the update files are .tar.gz, I came to realized that they are not actually gzip'ed but just tar'ed. If we want to make it safe, I'd recommend also adding the [gunzip-maybe](https://github.com/mafintosh/gunzip-maybe) that will check if a files is actually gzip'ed before attempting to extract it.

**Note 2:** I've managed to test as far as to check for an update, however my SP600 is already using the latest available firmware, so I **have no way of confirming that this is actually working as expected**, so I'm hoping someone with an SP600 can confirm if this works or not...

### CODE REVIEW

File: ota/salus.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    const images = (await axios.get(url)).data.versions;
+    const image = images.find((i) => i.model === modelId);
+    assert(image !== null, `No image available for modelId '${modelId}'`);
+    return {
+        fileVersion: parseInt(image.version, 16),
+        url: image.url.replace(/^http:\/\//, 'https://'),
+    };
+}
+
+async function untar(tarStream) {
```
Comment (PedroLamas) [inline ota/salus.js]: I wonder if we should move this to common, as I expect it might be useful elsewhere...

**Comment by tubededentifrice** [comment]
I'm wondering if you couldn't test it by just changing the JSON files (both the firmware list and the networkinfo.json file within the firmware tar you already have)

**Comment by PedroLamas** [comment]
Good question... in the end, we are talking about flashing the same firmware that the plug already has!

Not sure one should even try that as the update might just be patch instructions and not a full image to replace the old one.

**Comment by tubededentifrice** [comment]
Even if it's a patch, it should just work (replacing parts with actually the same parts) :D

**Comment by Koenkk** [comment]
@PedroLamas I expect it to be the full image, try forcing the update by e.g. removing the file version check.

**Comment by PedroLamas** [comment]
Fair enough, I'll do the test later today when I'm back home and report my findings!

**Comment by PedroLamas** [comment]
**SUCCESS!!**

```
zigbee2mqtt:info  2020-02-18 20:27:54: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Updating 'SmartPlug6' to latest firmware","meta":{"status":"update_in_progress","device":"SmartPlug6"}}'
zigbee2mqtt:info  2020-02-18 20:27:56: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 0%","meta":{"status":"update_progress","device":"SmartPlug6","progress":0}}'
zigbee2mqtt:info  2020-02-18 20:28:26: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 3.5%, +- 14 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":3.5}}'
zigbee2mqtt:info  2020-02-18 20:28:57: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 6.68%, +- 14 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":6.68}}'
zigbee2mqtt:info  2020-02-18 20:29:27: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 10.14%, +- 13 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":10.14}}'
zigbee2mqtt:info  2020-02-18 20:29:57: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 13.61%, +- 13 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":13.61}}'
zigbee2mqtt:info  2020-02-18 20:30:27: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 17.04%, +- 12 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":17.04}}'
zigbee2mqtt:info  2020-02-18 20:30:57: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 20.38%, +- 12 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":20.38}}'
zigbee2mqtt:info  2020-02-18 20:31:27: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 23.75%, +- 11 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":23.75}}'
zigbee2mqtt:info  2020-02-18 20:31:57: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 27.12%, +- 11 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":27.12}}'
zigbee2mqtt:info  2020-02-18 20:32:27: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 30.49%, +- 10 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":30.49}}'
zigbee2mqtt:info  2020-02-18 20:32:57: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 33.86%, +- 10 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":33.86}}'
zigbee2mqtt:info  2020-02-18 20:33:27: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 37.23%, +- 9 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":37.23}}'
zigbee2mqtt:info  2020-02-18 20:33:57: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 40.67%, +- 9 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":40.67}}'
zigbee2mqtt:info  2020-02-18 20:34:28: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 44.17%, +- 8 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":44.17}}'
zigbee2mqtt:info  2020-02-18 20:34:58: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 47.67%, +- 8 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":47.67}}'
zigbee2mqtt:info  2020-02-18 20:35:28: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 50.84%, +- 7 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":50.84}}'
zigbee2mqtt:info  2020-02-18 20:35:58: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 54.34%, +- 7 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":54.34}}'
zigbee2mqtt:info  2020-02-18 20:36:28: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 57.78%, +- 6 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":57.78}}'
zigbee2mqtt:info  2020-02-18 20:36:58: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 61.02%, +- 6 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":61.02}}'
zigbee2mqtt:info  2020-02-18 20:37:28: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 64.42%, +- 5 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":64.42}}'
zigbee2mqtt:info  2020-02-18 20:37:58: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 67.82%, +- 5 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":67.82}}'
zigbee2mqtt:info  2020-02-18 20:38:29: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 71.22%, +- 4 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":71.22}}'
zigbee2mqtt:info  2020-02-18 20:38:59: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 74.63%, +- 4 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":74.63}}'
zigbee2mqtt:info  2020-02-18 20:39:29: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 78.03%, +- 3 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":78.03}}'
zigbee2mqtt:info  2020-02-18 20:39:59: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 81.46%, +- 3 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":81.46}}'
zigbee2mqtt:info  2020-02-18 20:40:29: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 84.93%, +- 2 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":84.93}}'
zigbee2mqtt:info  2020-02-18 20:40:59: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 88.43%, +- 2 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":88.43}}'
zigbee2mqtt:info  2020-02-18 20:41:29: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 91.9%, +- 1 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":91.9}}'
zigbee2mqtt:info  2020-02-18 20:41:59: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 95.36%, +- 1 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":95.36}}'
zigbee2mqtt:info  2020-02-18 20:42:30: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 98.8%, +- 0 minutes remaining","meta":{"status":"update_progress","device":"SmartPlug6","progress":98.8}}'
zigbee2mqtt:info  2020-02-18 20:43:14: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Update of 'SmartPlug6' at 100%","meta":{"status":"update_progress","device":"SmartPlug6","progress":100}}'
zigbee2mqtt:info  2020-02-18 20:43:14: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"ota_update","message":"Finished update of 'SmartPlug6', from '{\"dateCode\":\"20170220\"}' to '{\"dateCode\":\"20170220\"}'","meta":{"status":"update_succeeded","device":"SmartPlug6","from":{"dateCode":"20170220"},"to":{"dateCode":"20170220"}}}
```

**Comment by Koenkk** [comment]
Awesome! Last thing: could you port the (small) change that I made in `tradfri.js` here: https://github.com/Koenkk/zigbee-herdsman-converters/commit/a6facb8c9b3b2cb62b92e2faa797dd6e1cfc7145 ?

**Comment by PedroLamas** [comment]
Done! From what I could see, it was just adding the `requestPayload` bit!

**Comment by Koenkk** [comment]
Awesome!! 🏅

**Comment by lwfitzgerald** [comment]
@PedroLamas Sorry for the late reply! I can also report success updating my SP600 using 1.11.0. This was from an older firmware version to the latest.

**Pull request merged**